src/main/java/ch/njol/skript/expressions/ExprFurnaceSlot.java

```diff
@@ -18,178 +18,233 @@
+	"set the block's ore slot to 64 iron ore",
+	"give the result of the block to the player",
+	"clear the result slot of the block"
+})
 @Events({"smelt", "fuel burn"})
-public class ExprFurnaceSlot extends PropertyExpression<Block, Slot> {
-	private final static int ORE = 0, FUEL = 1, RESULT = 2;
-	private final static String[] slotNames = {"ore", "fuel", "result"};
+@Since("1.0, INSERT VERSION (syntax rework)")
+public class ExprFurnaceSlot extends SimpleExpression<Slot> {
+
+	private static final int ORE = 0, FUEL = 1, RESULT = 2;
 	
 	static {
 		Skript.registerExpression(ExprFurnaceSlot.class, Slot.class, ExpressionType.PROPERTY,
-				"(" + FUEL + "¦fuel|" + RESULT + "¦result) [slot]",
-				"(" + ORE + "¦ore|" + FUEL + "¦fuel|" + RESULT + "¦result)[s] [slot[s]] of %blocks%",
-				"%blocks%'[s] (" + ORE + "¦ore|" + FUEL + "¦fuel|" + RESULT + "¦result)[s] [slot[s]]");
+				"[the] (0:ore slot|1:fuel slot|2:result [5:slot])",
+				"[the] (0:ore|1:fuel|2:result) slot[s] of %blocks%",
+				"%blocks%'[s] (0:ore|1:fuel|2:result) slot[s]"
+		);
 	}
-	
-	int slot;
-	boolean isEvent;
-	
-	@SuppressWarnings({"unchecked", "null"})
+
+	@Nullable
+	private Expression<Block> blocks;
+	private boolean isEvent;
+	private boolean isResultSlot;
+	private int slot;
+
 	@Override
-	public boolean init(final Expression<?>[] exprs, final int matchedPattern, final Kleenean isDelayed, final ParseResult parseResult) {
+	@SuppressWarnings("unchecked")
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
 		isEvent = matchedPattern == 0;
+		if (!isEvent)
+			blocks = (Expression<Block>) exprs[0];
+
 		slot = parseResult.mark;
-		if (isEvent && slot == RESULT && !getParser().isCurrentEvent(FurnaceSmeltEvent.class)) {
-			Skript.error("Cannot use 'result slot' outside a fuel smelt event.");
+		isResultSlot = slot == 7;
+		if (isResultSlot)
+			slot = RESULT;
+
+		if (isEvent && (slot == ORE || slot == RESULT) && !getParser().isCurrentEvent(FurnaceSmeltEvent.class)) {
+			Skript.error("Cannot use 'result slot' or 'ore slot' outside an ore smelt event.");
 			return false;
 		} else if (isEvent && slot == FUEL && !getParser().isCurrentEvent(FurnaceBurnEvent.class)) {
 			Skript.error("Cannot use 'fuel slot' outside a fuel burn event.");
 			return false;
 		}
-		if (!isEvent)
-			setExpr((Expression<Block>) exprs[0]);
+
 		return true;
 	}
+
+	@Override
+	@Nullable
+	protected Slot[] get(Event event) {
+		Block[] blocks;
+		if (isEvent) {
+			blocks = new Block[1];
+			if (event instanceof FurnaceSmeltEvent) {
+				blocks[0] = ((FurnaceSmeltEvent) event).getBlock();
+			} else if (event instanceof FurnaceBurnEvent) {
+				blocks[0] = ((FurnaceBurnEvent) event).getBlock();
+			} else {
+				return new Slot[0];
+			}
+		} else {
+			assert this.blocks != null;
+			blocks = this.blocks.getArray(event);
+		}
+
+		for (Block block : blocks) {
+			if (!ExprBurnCookTime.anyFurnace.isOfType(block))
+				return new Slot[0];
```

Comment:
Should this not `continue` instead? A block that isn't a furnace should just return null for itself, not cause the entire expression to return null in my opinion.